src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs

```diff
@@ -166,34 +500,77 @@ IOperation statement
             CurrentBasicBlock.AddStatement(statement);
         }
 
-        private void AppendNewBlock(BasicBlock block)
+        private void AppendNewBlock(BasicBlock block, bool linkToPrevious = true)
         {
             Debug.Assert(block != null);
-            BasicBlock prevBlock = _blocks.Last();
 
-            if (prevBlock.Next == null)
+            if (linkToPrevious)
             {
-                LinkBlocks(prevBlock, block);
+                BasicBlock prevBlock = _blocks.Last();
+
+                if (prevBlock.InternalNext.Destination == null)
+                {
+                    LinkBlocks(prevBlock, block);
+                }
             }
 
+            block.Ordinal = _blocks.Count;
             _blocks.Add(block);
             _currentBasicBlock = block;
+            _currentRegion.Add(block);
+            _regionMap.Add(block, _currentRegion);
+        }
+
+        private void EnterRegion(RegionBuilder region)
+        {
+            Debug.Assert(_currentRegion == region.Enclosing);
+            _currentRegion = region;
+            _currentBasicBlock = null;
+        }
+
+        private void LeaveRegion(RegionBuilder region)
+        {
+            Debug.Assert(_currentRegion == region);
+
+            // Ensure there is at least one block in the region
+            if (region.IsEmpty)
+            {
+                _ = CurrentBasicBlock;
```

Comment:
The fact that we're relying on a side-effect of this property here smells, and is completely nonobvious. I'd rather we duplicate the AddBlock call than do this.